Allow users to configure `host`, `port` and `timeout` for build in TCP adapters

RemcoSmitsDev opened the pull request:
This PR enhances user configurability for built-in adapters utilizing TCP transport by allowing customization of `host`, `port`, and `timeout` settings. Additionally, we've implemented a feature where, if no **port** is specified, the system will automatically attempt to locate an **available port** on the provided **host**.

TODO:
- [x] Update or remove the cached binaries functionality, as the host/port information is no longer stored within the DebugAdapterBinary

RemcoSmitsDev committed:
Remove space in cargo.toml

RemcoSmitsDev committed:
Add TCPHost for Javascript and PHP adapter

RemcoSmitsDev committed:
Allow falling back to first open port

RemcoSmitsDev committed:
Add some more docs to TCPHost

Anthony-Eid commented on crates/dap/src/transport.rs:
```diff
@@ -361,27 +361,38 @@ pub trait Transport: 'static + Send + Sync {
-                .ok()?
-                .port(),
-        )
+    pub async fn port(host: &TCPHost) -> Result<u16> {
```
I might be wrong, but we're binding an open port given to us by the OS. Then we telling a debug adapter to connect to that port that's already being used, couldn't that cause an error in some cases?

RemcoSmitsDev replied:
No, we are binding here, but that does not mean we picked the spot for the debug adapter to starts it's tcp server for that port.

Anthony-Eid commented on crates/task/src/debug_format.rs:
```diff
@@ -16,13 +16,26 @@ impl Default for DebugConnectionType {
+    /// The max amount of time in milliseconds to connect to a tcp DAP before returning an error
+    ///
+    /// Default: 2000ms
     pub timeout: Option<u64>,
```
We should probably make timeout a user setting too

RemcoSmitsDev replied:
Should that override evrything or should that be the default/fallback value that is now 2000ms?

RemcoSmitsDev committed:
Fix cached binaries prevent from having multiple debug sessions for one adapters

This was an issue, because we stored the port arg inside the DebugAdapterBinary which was cached so could not change anymore.

Co-authored-by: Anthony Eid <hello@anthonyeid.me>

RemcoSmitsDev committed:
Merge branch 'debugger' into allow-dynamic-tcp-ports

RemcoSmitsDev committed:
Add default setting for tcp timeout

Co-Authored-By: Anthony Eid <56899983+Anthony-Eid@users.noreply.github.com>
